mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -217,6 +218,7 @@ async fn block_header_utxo_loop<T: UtxoCommonOps>(
 ) {
     while let Some(arc) = weak.upgrade() {
         let coin = constructor(arc);
+        let ticker = coin.as_ref().conf.ticker.as_str();
```

Comment:
This what I meant by how `block_header_utxo_loop` can be simplified

``` rust
const BLOCK_HEADERS_ACCEPTABLE_CHUNK_SIZE: u64 = 500;
const ELECTRUM_MAX_CHUNK_SIZE: u64 = 2016;

async fn block_header_utxo_loop<T: UtxoCommonOps>(
    weak: UtxoWeak,
    constructor: impl Fn(UtxoArc) -> T,
    mut sync_status_loop_handle: UtxoSyncStatusLoopHandle,
    mut last_block_height: u64,
) {
    let mut chunk_size = ELECTRUM_MAX_CHUNK_SIZE;
    while let Some(arc) = weak.upgrade() {
        let coin = constructor(arc);
        let ticker = coin.as_ref().conf.ticker.as_str();
        let client = match &coin.as_ref().rpc_client {
            UtxoRpcClientEnum::Native(_) => break,
            UtxoRpcClientEnum::Electrum(client) => client,
        };

        let storage = client.block_headers_storage();
        let from_block_height = match storage.get_last_block_height().await {
            Ok(h) => h,
            Err(e) => {
                error!("Error {} on getting the height of the last stored header in DB!", e);
                sync_status_loop_handle.notify_on_temp_error(e.to_string());
                Timer::sleep(10.).await;
                continue;
            },
        };

        // Todo: Add code for the case if a chain reorganization happens
        if from_block_height == last_block_height {
            sync_status_loop_handle.notify_sync_finished(last_block_height);
            last_block_height = match client.get_block_count().compat().await {
                Ok(h) => h,
                Err(e) => {
                    error!("Error {} on getting the height of the latest block from rpc!", e);
                    sync_status_loop_handle.notify_on_temp_error(e.to_string());
                    Timer::sleep(10.0).await;
                    continue;
                },
            };
            Timer::sleep(BLOCK_HEADERS_LOOP_INTERVAL).await;
            continue;
        }

        let to_block_height = from_block_height + chunk_size;
        sync_status_loop_handle.notify_blocks_headers_sync_status(from_block_height + 1, to_block_height);

        let (block_registry, block_headers, last_retrieved_height) = match client
            .retrieve_headers(from_block_height + 1, to_block_height)
            .compat()
            .await
        {
            Ok(res) => res,
            Err(error) => {
                if error.get_inner().is_network_error() {
                    log!("Network Error: Will try fetching block headers again after 10 secs");
                    sync_status_loop_handle.notify_on_temp_error(error.to_string());
                    Timer::sleep(10.).await;
                    continue;
                };

                // If electrum returns response too large error", we will request the headers again in chunks of BLOCK_HEADERS_ACCEPTABLE_CHUNK_SIZE instead.
                if error.get_inner().is_response_too_large() {
                    chunk_size = BLOCK_HEADERS_ACCEPTABLE_CHUNK_SIZE;
                    continue;
                }

                error!("Error {} on retrieving the latest headers from rpc!", error);
                sync_status_loop_handle.notify_on_permanent_error(error.to_string());
                break;
            },
        };

        // Validate retrieved block headers
        if let Some(params) = &coin.as_ref().conf.block_headers_verification_params {
            if let Err(e) = validate_headers(ticker, from_block_height, block_headers, storage, params).await {
                error!("Error {} on validating the latest headers!", e);
                // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
                sync_status_loop_handle.notify_on_permanent_error(e.to_string());
                break;
            }
        };

        ok_or_continue_after_sleep!(
            storage.add_block_headers_to_storage(block_registry).await,
            BLOCK_HEADERS_LOOP_INTERVAL
        );
    }
}
```
A few notes:
- You will need to pass `last_block_height` to the function
- `get_block_count` calls is reduced a lot now, this will reduce the electrum session cost https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html#envvar-COST_SOFT_LIMIT
- We can even not use `BLOCK_HEADERS_ACCEPTABLE_CHUNK_SIZE` and reduce the `chunk_size` by a 100 when we get `response too large` error, this will be beneficial in reducing the number of requests, please apply this to the code example above if you think it's a good idea. We will be able to get the optimal `chunk_size` in this case :)
- `last_retrieved_height` is not needed anymore, please refactor `retrieve_headers` function accordingly.

Reply:
all notes checked and updated.